Review

os-max opened the pull request:
Added full functionality for review

os-max committed:
feat all form actions and input handling

os-max committed:
feat full functionality. fixed checkbox not working after reset. fixed minor fixes to styling

os-max committed:
feat Form closes on click outside form container

mikebars1995 commented on src/index.tsx:
```diff
@@ -13,19 +16,36 @@ const domNode = document.getElementById('root') as HTMLDivElement;
 const root = createRoot(domNode);
 
 const App = () => {
+	const [resetKey, setResetKey] = useState(false);
+	const [appState, setAppState] = useState<IFormState>(defaultArticleState);
+
+	function handleSubmit(event: FormEvent, state: typeof appState) {
+		event.preventDefault();
+		setAppState({ ...state });
+	}
+
+	function handleReset() {
+		setAppState({ ...defaultArticleState });
+		setResetKey(!resetKey);
+	}
```
**Нужно исправить**


Состояние формы должно быть в компоненте `ArticleParamsForm`, а не `App`


Каждый компонент сам должен отвечать за хранение и обработку своего состояния. Использовать подъем состояния формы в родительский компонент не имеет смысла для нашего приложения. Состояние формы нужно описать в самом компонент `ArticleParamsForm`, а не передавать его из App, т.к. это состояние описывает именно `ArticleParamsForm` и больше вне самого компонента `ArticleParamsForm` данные о текущем состоянии формы нигде не нужны.  

В состоянии `App` должны хранится только данные о текущем отображении статьи.

mikebars1995 commented on src/index.tsx:
```diff
@@ -13,19 +16,36 @@ const domNode = document.getElementById('root') as HTMLDivElement;
 
 const App = () => {
+	const [resetKey, setResetKey] = useState(false);
+	const [appState, setAppState] = useState<IFormState>(defaultArticleState);
```
**Нужно исправить**

По чеклисту - “имена переменных — существительные, отражающие то, что в них хранится”. Это относится в том числе и к стейту. Названия, например `state appState`, недостаточно информативны, должны быть использованы более осмысленные имена - `articleState` `formState` и т.д.

mikebars1995 commented on src/index.tsx:
```diff
@@ -13,19 +16,36 @@ const domNode = document.getElementById('root') as HTMLDivElement;
+	}
+
 	return (
 		<div
```
**Можно лучше**

* По семантике нужно помещать весь главный контент страницы **после** хэдера в тег `main`.

mikebars1995 commented on src/index.tsx:
```diff
@@ -13,19 +16,36 @@ const domNode = document.getElementById('root') as HTMLDivElement;
 const root = createRoot(domNode);
 
 const App = () => {
```
**Можно лучше**

* Лучше выделить компонент  `App` в отдельный файл, чтобы не смешивать все в `index.js`

mikebars1995 commented on src/index.tsx:
```diff
@@ -13,19 +16,36 @@ const domNode = document.getElementById('root') as HTMLDivElement;
+
 	return (
 		<div
 			className={clsx(styles.main)}
```
**Можно лучше**

* `clsx` можно не использовать, если только **один** селектор передается

mikebars1995 commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,20 +1,152 @@
+	onReset,
+	resetKey,
+}: FormProps) => {
+	const [isOpen, setIsOpen] = useState<boolean>(false);
```
**Нужно исправить**

* `[isOpen, setIsOpen]` нужно назвать `[isMenuOpen, setIsMenuOpen]`, чтобы было понятно, что тут открываете

mikebars1995 commented on src/components/article-params-form/hooks/useClickOutside.tsx:
```diff
@@ -0,0 +1,28 @@
+	ref,
+	callBack,
+}: IUseClickOutside) {
+	useEffect(() => {
```
**Нужно исправить**


* Нужно **останавливать эффект**, если форма закрыта, чтобы не навешивать обработчик 

```jsx
if (!isOpen) return;

```

mikebars1995 reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на *гитхабе* все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная


Что сделано отлично:

* Хорошо реализовали форму из предоставленных компонентов
* Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку. 
* Отлично, что нет типов `any`
* Отлично типизировали проект


Что нужно исправить:

* Состояние формы должно быть в компоненте `ArticleParamsForm`, а не `App`
* По чеклисту - “имена переменных — существительные, отражающие то, что в них хранится”. Это относится в том числе и к стейту. 
* `[isOpen, setIsOpen]` нужно назвать `[isMenuOpen, setIsMenuOpen]`, чтобы было понятно, что тут открываете	
* Нужно **останавливать эффект**, если форма закрыта, чтобы не навешивать обработчик 

Что можно улучшить:


* По семантике нужно помещать весь главный контент страницы **после** хэдера в тег `main`.
* Лучше выделить компонент  `App` в отдельный файл, чтобы не смешивать все в `index.js`
* `clsx` можно не использовать, если только **один** селектор передается


Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте работоспособность проекта и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

**Комментарии `Можно лучше` не обязательны к исправлению прямо сейчас, это рекомендации**

Удачного рефакторинга кода.

os-max committed:
App component moved to it's own file
Fixed naming of states
Fixed useClickOutside now checks if form is open before adding event listner
Fixed minor fixes to styling

mikebars1995 reviewed (commented):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
